plugins/outputs/otlp/otlp.go

```diff
@@ -0,0 +1,379 @@
+package otlp
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sort"
+	"time"
+
+	// TODO: this should be simplified
+	// Imports the OTLP client from the prometheus sidecar
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+	metricsService "go.opentelemetry.io/proto/otlp/collector/metrics/v1"
+	otlpcommonpb "go.opentelemetry.io/proto/otlp/common/v1"
+	otlpmetricpb "go.opentelemetry.io/proto/otlp/metrics/v1"
+	otlpresourcepb "go.opentelemetry.io/proto/otlp/resource/v1"
+	"google.golang.org/grpc/metadata"
+)
+
+// OTLP is the OpenTelemetry Protocol config info.
+type OTLP struct {
+	Endpoint   string            `toml:"endpoint"`
+	Timeout    string            `toml:"timeout"`
+	Headers    map[string]string `toml:"headers"`
+	Attributes map[string]string `toml:"attributes"`
+
+	Namespace string
+	Log       telegraf.Logger
+
+	client       *Client
+	resourceTags []*telegraf.Tag
+	grpcTimeout  time.Duration
+}
+
+const (
+	// QuotaLabelsPerMetricDescriptor is the limit
+	// to labels (tags) per metric descriptor.
+	QuotaLabelsPerMetricDescriptor = 10
+	// QuotaStringLengthForLabelKey is the limit
+	// to string length for label key.
+	QuotaStringLengthForLabelKey = 100
+	// QuotaStringLengthForLabelValue is the limit
+	// to string length for label value.
+	QuotaStringLengthForLabelValue = 1024
+
+	// StartTime for cumulative metrics.
+	StartTime = int64(1)
+	// MaxInt is the max int64 value.
+	MaxInt = int(^uint(0) >> 1)
+
+	defaultEndpoint            = "http://localhost:4317"
+	defaultTimeout             = time.Second * 60
+	instrumentationLibraryName = "Telegraf"
+)
+
+var sampleConfig = `
+  ## OpenTelemetry endpoint
+  endpoint = "http://localhost:4317"
+
+  ## Timeout used when sending data over grpc
+  timeout = "10s"
+
+  # Additional resource attributes
+  [outputs.otlp.attributes]
+  	"service.name" = "demo"
+
+  # Additional grpc metadata
+  [outputs.otlp.headers]
+    key1 = "value1"
+
+`
+
+// Connect initiates the primary connection to the OTLP endpoint.
+func (o *OTLP) Connect() error {
+	if o.Endpoint == "" {
+		o.Endpoint = defaultEndpoint
+	}
+	endpoint, err := url.Parse(o.Endpoint)
+	if err != nil {
+		return fmt.Errorf("invalid endpoint configured")
+	}
+
+	if o.Timeout == "" {
+		o.grpcTimeout = defaultTimeout
+	} else {
+		o.grpcTimeout, err = time.ParseDuration(o.Timeout)
+		if err != nil {
+			return fmt.Errorf("invalid timeout configured")
+		}
+	}
+
+	for k, v := range o.Attributes {
+		o.resourceTags = append(o.resourceTags, &telegraf.Tag{Key: k, Value: v})
+	}
+
+	if o.Headers == nil {
+		o.Headers = make(map[string]string, 1)
+	}
+
+	o.Headers["telemetry-reporting-agent"] = fmt.Sprint(
+		"telegraf/",
+		internal.Version(),
+	)
+
```

Comment:
I think most of this should move to the `Init()` function. Even if it makes no difference on the program flow it makes clear what is initialization and what is connection handling.

Reply:
I moved most of this to `Init`, thanks for the tip, i had no idea `Init` even existed :)